File: apps/android/app/src/main/java/org/janark/app/ui/profile/ProfileScreen.kt
package org.janark.app.ui.profile

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import org.janark.app.core.Outcome
import org.janark.app.core.formatDate
import org.janark.app.core.hrefToRoute
import org.janark.app.core.indianCount
import org.janark.app.data.net.ProfilePostItem
import org.janark.app.data.net.ProfileResponse
import org.janark.app.ui.LocalApp
import org.janark.app.ui.auth.LoginScreen
import org.janark.app.ui.components.AmberButton
import org.janark.app.ui.components.EmptyState
import org.janark.app.ui.components.ErrorState
import org.janark.app.ui.components.GhostButton
import org.janark.app.ui.components.KindPill
import org.janark.app.ui.components.LoadingBox
import org.janark.app.ui.components.SectionLabel
import org.janark.app.ui.components.SurfaceCard
import org.janark.app.ui.nav.Routes
import org.janark.app.ui.theme.JanarkTheme
import org.janark.app.ui.theme.toC

@Composable
fun ProfileGate(anonId: String?, onNavigate: (String) -> Unit) {
    val app = LocalApp.current
    val session by app.session.uiSession.collectAsState(initial = org.janark.app.data.session.UiSession())
    val id = anonId ?: session.anonId
    if (id.isNullOrBlank()) {
        LoginScreen(reason = "see your profile", onSuccess = {
            session.anonId?.let { onNavigate(Routes.profile(it)) }
        })
    } else {
        ProfileScreen(id, onNavigate)
    }
}

@Composable
fun ProfileScreen(anonId: String, onNavigate: (String) -> Unit) {
    val app = LocalApp.current
    val session by app.session.uiSession.collectAsState(initial = org.janark.app.data.session.UiSession())
    var data by remember { mutableStateOf<ProfileResponse?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var loading by remember { mutableStateOf(true) }
    val scope = rememberCoroutineScope()
    val c = JanarkTheme.colors
    fun load() {
        loading = true
        scope.launch {
            when (val r = app.repository.profile(anonId)) {
                is Outcome.Ok -> data = r.value
                is Outcome.Err -> error = r.message
            }
            loading = false
        }
    }
    LaunchedEffect(anonId) { load() }
    when {
        loading && data == null -> LoadingBox()
        error != null && data == null -> ErrorState(error!!) { load() }
        data == null -> EmptyState("Profile not found")
        else -> {
            val p = data!!
            val items = buildList {
                addAll(p.posts.demands.map { it to "Petition" })
                addAll(p.posts.reports.map { it to "Report" })
                addAll(p.posts.discussions.map { it to "Discussion" })
                addAll(p.posts.notices.map { it to "Notice" })
                addAll(p.posts.shares.map { it to "Share" })
                addAll(p.posts.memes.map { it to "Post" })
            }
            LazyColumn(Modifier.fillMaxSize(), contentPadding = PaddingValues(20.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
                item {
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        Text(p.profile?.label ?: anonId, fontFamily = FontFamily.Serif, fontSize = 28.sp, fontWeight = FontWeight.SemiBold, color = c.foreground.toC())
                        Text(anonId, fontFamily = FontFamily.Monospace, color = c.muted.toC())
                        Text("Member since ${formatDate(p.counts.memberSince ?: p.profile?.memberSince)}", color = c.muted.toC(), fontSize = 13.sp)
                        Text(
                            "${indianCount(p.counts.followers)} followers · ${indianCount(p.counts.following)} following · ${indianCount(p.counts.posts)} posts",
                            color = c.foreground.toC(),
                        )
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            GhostButton("Followers") { onNavigate(Routes.followList(anonId, "followers")) }
                            GhostButton("Following") { onNavigate(Routes.followList(anonId, "following")) }
                        }
                        when {
                            p.follow.isSelf -> Text("This is you. Your phone number is never shown.", color = c.muted.toC(), fontSize = 13.sp)
                            session.authenticated -> AmberButton(if (p.follow.viewerFollows) "Unfollow" else "Follow", onClick = {
                                scope.launch {
                                    app.repository.toggleFollow(anonId, if (p.follow.viewerFollows) "unfollow" else "follow")
                                    load()
                                }
                            })
                            else -> AmberButton("Follow", onClick = {
                                app.auth.require("follow") {
                                    scope.launch {
                                        app.repository.toggleFollow(anonId, "follow")
                                        load()
                                    }
                                }
                            })
                        }
                        SectionLabel("Posts")
                    }
                }
                if (items.isEmpty()) {
                    item { Text("No public posts yet.", color = c.muted.toC()) }
                }
                items(items, key = { (item, kind) -> "${kind}-${item.id}-${item.href}" }) { (item, kind) ->
                    ProfileRow(item, kind, onNavigate)
                }
                item {
                    SectionLabel("Activity")
                    val acts = p.reactions.demandSupports + p.reactions.proposalVotes + p.reactions.reportVotes + p.reactions.noticeSignatures
                    if (acts.isEmpty()) Text("No public reactions listed.", color = c.muted.toC())
                    acts.take(20).forEach { a ->
                        SurfaceCard(onClick = { hrefToRoute(a.href)?.let(onNavigate) }) {
                            Text(a.title ?: "Activity", color = c.foreground.toC())
                            Text(listOfNotNull(a.choice, a.reaction, a.ask).joinToString(" · "), color = c.muted.toC(), fontSize = 12.sp)
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun ProfileRow(item: ProfilePostItem, kind: String, onNavigate: (String) -> Unit) {
    val c = JanarkTheme.colors
    SurfaceCard(onClick = { hrefToRoute(item.href)?.let(onNavigate) }) {
        KindPill(kind, c.amber)
        Text(item.title ?: item.body ?: item.ask ?: item.caption ?: "Post", color = c.foreground.toC(), fontWeight = FontWeight.SemiBold, modifier = Modifier.padding(top = 6.dp))
    }
}

@Composable
fun FollowListScreen(anonId: String, list: String, onNavigate: (String) -> Unit) {
    val app = LocalApp.current
    var names by remember { mutableStateOf(listOf<org.janark.app.data.net.FollowPersonDto>()) }
    var error by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    LaunchedEffect(anonId, list) {
        when (val r = app.repository.followStatus(anonId, list)) {
            is Outcome.Ok -> names = r.value.people
            is Outcome.Err -> error = r.message
        }
    }
    val c = JanarkTheme.colors
    Column(Modifier.fillMaxSize().padding(20.dp), verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Text(list.replaceFirstChar { it.uppercase() }, fontFamily = FontFamily.Serif, fontSize = 24.sp, color = c.foreground.toC())
        error?.let { Text(it, color = c.danger.toC()) }
        names.forEach { p ->
            SurfaceCard(onClick = { onNavigate(Routes.profile(p.anonId)) }) {
                Text(p.label ?: p.anonId, color = c.foreground.toC(), fontWeight = FontWeight.SemiBold)
                Text(p.anonId, fontFamily = FontFamily.Monospace, color = c.muted.toC(), fontSize = 12.sp)
            }
        }
    }
}

File: apps/ios/Janark/UI/Profile/ProfileScreen.swift
import SwiftUI

struct ProfileGate: View {
    @EnvironmentObject private var app: AppContainer
    let anonId: String?
    let onNavigate: (String) -> Void

    var body: some View {
        let id = anonId ?? app.session.uiSession.anonId
        if let id, !id.isEmpty {
            ProfileScreen(anonId: id, onNavigate: onNavigate)
        } else {
            LoginScreen(reason: "see your profile", onSuccess: {
                if let anon = app.session.uiSession.anonId {
                    onNavigate(Routes.profile(anon))
                }
            })
        }
    }
}

struct ProfileScreen: View {
    @EnvironmentObject private var app: AppContainer
    @Environment(\.janarkColors) private var c
    let anonId: String
    let onNavigate: (String) -> Void

    @State private var data: ProfileResponse?
    @State private var error: String?
    @State private var loading = true

    var body: some View {
        Group {
            if loading && data == nil {
                LoadingView()
            } else if let error, data == nil {
                ErrorView(message: error) { Task { await load() } }
            } else if let p = data {
                profileBody(p)
            } else {
                EmptyStateView(title: "Profile not found")
            }
        }
        .task(id: anonId) { await load() }
    }

    private func profileBody(_ p: ProfileResponse) -> some View {
        let items: [(ProfilePostItem, String)] =
            p.posts.demands.map { ($0, "Petition") }
            + p.posts.reports.map { ($0, "Report") }
            + p.posts.discussions.map { ($0, "Discussion") }
            + p.posts.notices.map { ($0, "Notice") }
            + p.posts.shares.map { ($0, "Share") }
            + p.posts.memes.map { ($0, "Post") }

        return ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                Text(p.profile?.label ?? anonId)
                    .font(.system(.largeTitle, design: .serif).weight(.semibold))
                    .foregroundStyle(c.foreground)
                Text(anonId)
                    .font(.system(.body, design: .monospaced))
                    .foregroundStyle(c.muted)
                Text("Member since \(formatDate(p.counts.memberSince ?? p.profile?.memberSince))")
                    .font(.subheadline)
                    .foregroundStyle(c.muted)
                Text("\(indianCount(p.counts.followers)) followers · \(indianCount(p.counts.following)) following · \(indianCount(p.counts.posts)) posts")
                    .foregroundStyle(c.foreground)

                HStack(spacing: 8) {
                    GhostButton("Followers") { onNavigate(Routes.followList(anonId, "followers")) }
                    GhostButton("Following") { onNavigate(Routes.followList(anonId, "following")) }
                }

                followControls(p)

                SectionLabel("Posts")
                if items.isEmpty {
                    Text("No public posts yet.").foregroundStyle(c.muted)
                } else {
                    ForEach(Array(items.enumerated()), id: \.offset) { _, pair in
                        profileRow(pair.0, kind: pair.1)
                    }
                }

                SectionLabel("Activity")
                let acts = p.reactions.demandSupports
                    + p.reactions.proposalVotes
                    + p.reactions.reportVotes
                    + p.reactions.noticeSignatures
                if acts.isEmpty {
                    Text("No public reactions listed.").foregroundStyle(c.muted)
                } else {
                    ForEach(Array(acts.prefix(20).enumerated()), id: \.offset) { _, a in
                        SurfaceCard(onTap: { hrefToRoute(a.href).map(onNavigate) }) {
                            Text(a.title ?? "Activity").foregroundStyle(c.foreground)
                            Text([a.choice, a.reaction, a.ask].compactMap { $0 }.joined(separator: " · "))
                                .font(.caption)
                                .foregroundStyle(c.muted)
                        }
                    }
                }
            }
            .padding(20)
        }
    }

    @ViewBuilder
    private func followControls(_ p: ProfileResponse) -> some View {
                if p.follow.isSelf {
                    Text("This is you. Your phone number is never shown.")
                        .font(.footnote)
                        .foregroundStyle(c.muted)
                } else if app.session.uiSession.authenticated {
            AmberButton(p.follow.viewerFollows ? "Unfollow" : "Follow") {
                Task {
                    _ = await app.repository.toggleFollow(
                        anonId: anonId,
                        action: p.follow.viewerFollows ? "unfollow" : "follow"
                    )
                    await load()
                }
            }
        } else {
            AmberButton("Follow") {
                app.auth.require("follow") {
                    Task {
                        _ = await app.repository.toggleFollow(anonId: anonId, action: "follow")
                        await load()
                    }
                }
            }
        }
    }

    private func profileRow(_ item: ProfilePostItem, kind: String) -> some View {
        SurfaceCard(onTap: { hrefToRoute(item.href).map(onNavigate) }) {
            KindPill(label: kind, color: c.amber)
            Text(item.title ?? item.body ?? item.ask ?? item.caption ?? "Post")
                .font(.headline)
                .foregroundStyle(c.foreground)
                .padding(.top, 6)
        }
    }

    private func load() async {
        loading = true
        switch await app.repository.profile(anonId: anonId) {
        case .ok(let value): data = value
        case .err(let message, _, _): error = message
        }
        loading = false
    }
}

struct FollowListScreen: View {
    @EnvironmentObject private var app: AppContainer
    @Environment(\.janarkColors) private var c
    let anonId: String
    let list: String
    let onNavigate: (String) -> Void

    @State private var names: [FollowPersonDto] = []
    @State private var error: String?

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 10) {
                Text(list.capitalized)
                    .font(.system(.title, design: .serif))
                    .foregroundStyle(c.foreground)
                if let error {
                    Text(error).foregroundStyle(c.danger)
                }
                ForEach(names, id: \.anonId) { p in
                    SurfaceCard(onTap: { onNavigate(Routes.profile(p.anonId)) }) {
                        Text(p.label ?? p.anonId)
                            .font(.headline)
                            .foregroundStyle(c.foreground)
                        Text(p.anonId)
                            .font(.system(.caption, design: .monospaced))
                            .foregroundStyle(c.muted)
                    }
                }
            }
            .padding(20)
        }
        .task(id: "\(anonId)-\(list)") {
            switch await app.repository.followStatus(anonId: anonId, list: list) {
            case .ok(let value): names = value.people
            case .err(let message, _, _): error = message
            }
        }
    }
}
